File: packages/core/src/api/BlockRegistry.ts
import { BlockCategory } from '../model/BlockCategory.js';
import { BlockDefinition } from '../model/BlockDefinition.js';
import catalog from '../catalog/blocks.json' with { type: 'json' };

const blocks = catalog as BlockDefinition[];
const byType = new Map<string, BlockDefinition>(blocks.map((b) => [b.type, b]));

export function getBlock(type: string): BlockDefinition | undefined {
  return byType.get(type);
}

export function getAllBlocks(): BlockDefinition[] {
  return blocks;
}

export function getByCategory(category: BlockCategory): BlockDefinition[] {
  return blocks.filter((b) => b.category === category);
}

File: java/flowcamel-core/src/main/java/com/flowcamel/core/registry/BlockRegistry.java
package com.flowcamel.core.registry;

import com.flowcamel.core.catalog.CatalogJson;
import com.flowcamel.core.model.BlockCategory;
import com.flowcamel.core.model.BlockDefinition;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

public final class BlockRegistry {
  private static final List<BlockDefinition> BLOCKS = loadBlocks();
  private static final Map<String, BlockDefinition> BY_TYPE =
      BLOCKS.stream().collect(Collectors.toMap(b -> b.type, b -> b));

  private static List<BlockDefinition> loadBlocks() {
    BlockDefinition[] arr = CatalogJson.read("catalog/blocks.json", BlockDefinition[].class);
    return List.of(arr);
  }

  private BlockRegistry() {}

  public static Optional<BlockDefinition> getBlock(String type) {
    return Optional.ofNullable(BY_TYPE.get(type));
  }

  public static List<BlockDefinition> getAllBlocks() {
    return BLOCKS;
  }

  public static List<BlockDefinition> getByCategory(BlockCategory category) {
    return BLOCKS.stream().filter(b -> b.category == category).toList();
  }
}
